keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/Forward.java

```diff
@@ -0,0 +1,110 @@
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    @Override
+    public NetworkSamples getPosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom, int sampleCount) {
+        return generatePosteriorSamples(model, variablesToSampleFrom)
+            .generate(sampleCount);
+    }
+
+    @Override
+    public NetworkSamplesGenerator generatePosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        return new NetworkSamplesGenerator(setupSampler(model, variablesToSampleFrom), StatusBar::new);
+    }
+
+    private SamplingAlgorithm setupSampler(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        List<? extends Variable> latentVariables = model.getLatentVariables();
+        Preconditions.checkArgument(latentVariables.size() > 0, "Your model must contain latent variables in order to forward sample.");
+
+        Variable latent = latentVariables.get(0);
+        Preconditions.checkArgument(latent instanceof Vertex, "The Forward Sampler only works for Variables of type Vertex.");
+
+        BayesianNetwork network = new BayesianNetwork(((Vertex) latent).getConnectedGraph());
+
+        List<Vertex> observedVertices = network.getObservedVertices();
+        checkUpstreamOfObservedDoesNotContainProbabilistic(observedVertices);
+
+        List<Vertex> verticesToSampleFrom = variablesToSampleFrom.stream().map(v -> (Vertex) v).collect(Collectors.toList());
+        Set<Vertex> allUpstreamVertices = allUpstreamVertices(verticesToSampleFrom);
+
+        List<Vertex> sortedVertices = TopologicalSort.sort(allUpstreamVertices);
+        sortedVertices = removeNonProbabilisticVerticesBeforeTheFirstProbabilistic(sortedVertices);
+
+        return new ForwardSampler(verticesToSampleFrom, sortedVertices, random);
+    }
+
+    private List<Vertex> removeNonProbabilisticVerticesBeforeTheFirstProbabilistic(List<Vertex> topologicalVertices) {
+        List<Vertex> copy = new ArrayList<>(topologicalVertices);
+
+        for (Vertex vertex : topologicalVertices) {
+            if (vertex instanceof Probabilistic) {
+                break;
+            } else {
+                copy.remove(vertex);
+            }
+        }
+        return copy;
+    }
+
+    private Set<Vertex> allUpstreamVertices(List<Vertex> fromVertices) {
+        Set<Vertex> upstream = new HashSet<>();
+        for (Vertex vertex : fromVertices) {
+            LambdaSection upstreamLambdaSection = LambdaSection.getUpstreamLambdaSection(vertex, true);
+            Set<Vertex> vertexUpsteam = upstreamLambdaSection.getAllVertices();
+            upstream = Sets.union(upstream, vertexUpsteam);
+        }
+        return upstream;
+    }
+
+    private void checkUpstreamOfObservedDoesNotContainProbabilistic(List<Vertex> vertices) {
+        for (Vertex vertex : vertices) {
```

Comment:
Probably not a huge pain at the moment, but obviously this could cause some performance issues in some large graphs with lots of observed variables.  If so we'd probably want to enhance how the Lambda section stuff works.